app.py

```diff
@@ -89,9 +91,40 @@ def edit_forum_comment(post_id, comment_id):
 def login_signup():
     return render_template('login_signup.html', login_active=True)
 
+@app.route('/process_form', methods=['POST'])
+def process_form():
+    username = request.form.get('username')
+    hashed_password = request.form.get('hashed_password')
+    action = request.form.get('action')
+    
+    if not username or not hashed_password:
+        abort(400)
+
+    if action == 'Sign Up':
+        bcrypt_password = bcrypt.generate_password_hash(hashed_password).decode()
+        new_user = AppUser(username=username, hashed_password=bcrypt_password)
+        db.session.add(new_user)
+        db.session.commit()
+        return redirect('/login_signup.html')
+    elif action == 'Login':
+        username=request.form.get('username')
+        hashed_password=request.form.get('hashed_password')
+        if not username or not hashed_password:
+            abort (400)
+        existing_user = AppUser.query.filter_by(username=username).first_or_404()
+        if not existing_user:
+            abort(401)
+        if not bcrypt.check_password_hash(existing_user.password, hashed_password):
+            abort(401)
+        return redirect('/user_profile')
```

Comment:
username needs to be added to the session dict before redirecting so that the session check in userprofile() works